### ISSUE TITLE
[Bug] rsync_cleanup_after_native_cp_al() only works on directories [sf#24]

### ISSUE THREAD

**Issue by bebehei** [rsnapshot/sourceforge-issues#36]
**Reported by nlshep on 2010-03-08 07:03 UTC**
I've set up rsnapshot (version 1.3.1, with rsync version 3.0.6 on Gentoo Linux 32-bit)  to back up my personal documents tree.  When I first started everything worked fine.  At some point, however, backups stopped working, with the following error:
----------------------------------------------------------------------------
rsnapshot encountered an error! The program was invoked with these options:
/usr/bin/rsnapshot -Dc /etc/rsnapshot/nls.conf decminly 
----------------------------------------------------------------------------
ERROR: rsync_cleanup_after_native_cp_al() only works on directories
ERROR: Error! cp_al("/mnt/scratch/snapshots/nls/decminly.0/", "/mnt/scratch/snapshots/nls/decminly.1/")

decminly.0 and decminly.1 are indeed directories.  I tracked the problem down to a check on line 4691 (line number from the rsnapshot-program.pl file in the tarball), which determines whether something is a directory based on whether it has a trailing slash.  Because, in certain instances, rsnapshot creates and uses paths without trailing slashes (for example, line 3233), this check fails.  Wouldn't it be better to use the standard "-d some_path" for checking whether something is a directory, or is there some special reason the decision is made based on trailing slash here? If I change the condition to use the "-d" check, then everything works fine.  Attached is a patch which implements this change.

**Comment by bebehei** [comment]
**Commented by nlshep on 2010-03-08 07:03 UTC**
patch which changes from pattern matching to the "-d" condition to check whether something is a dir
Attached filefix_dir_test.patch:

```
--- rsnapshot-program.pl    2008-08-31 07:17:07.000000000 -0400
+++ rsnapshot-program-fix_dir_test.pl   2010-03-08 01:56:26.000000000 -0500
@@ -4688,7 +4688,7 @@
    if (!defined($dest)) { return(0); }

    # make sure this is directory to directory
-   if (($src !~ m/\/$/o) or ($dest !~ m/\/$/o)) {
+   if ((! -d $src) or (! -d $dest)) {
        print_err("rsync_cleanup_after_native_cp_al() only works on directories", 2);
        return (0);
    }
```

**Comment by bebehei** [comment]
**Commented by nlshep on 2010-03-08 15:53 UTC**
Actually, I just noticed that below the part that was pattern matching for a trailing slash, there are if statements that actually use the '-d' conditional.  My guess then is that you're looking for the trailing slash because rsync cares about the trailing slash on the source. Is that right?  It would seem to me that you don't need to do that for the destination, because rsync doesn't care about a trailing slash on the destination.

**Comment by tonyrein** [comment]
This conversation is old, but I just encountered this error on rsnapshot version 1.4.0, running on FreeBSD (in a FreeNAS "native" jail). The FreeBSD version is 9.3-RELEASE-p23, in case that matters.

I think I've fixed it. I installed the GNU coreutils package and defined cmd_cp as "/usr/local/bin/gcp" to use that cp command instead of the native FreeBSD one. I reran the rsnapshot -c rsnapshot.conf sync command and it completed without error.

I don't think this is directly relevant to the problems mentioned above, but I hope it helps someone.
